src/factor.h

```diff
@@ -63,10 +71,18 @@ class CompactFactor {
   inline double potential(const VariableInFactor vifs[],
                           const variable_value_t var_values[],
                           const variable_id_t vid,
-                          const variable_value_t proposal) const {
+                          const variable_value_t proposal,
+                          feature_value_t val = DEFAULT_FEATURE_VALUE) const {
+    // For boolean, this is stored in CompactFactor.value;
+    //     caller shouldn't set the val arg.
+    // For categorical, this is stored in Factor.factor_params.value.
+    //     caller should pass in via the val arg.
+    // TODO: better data structures than the factor_params duct tape...
+    // See also: CompactFactorGraph::potential
+    val = (func_id == FUNC_AND_CATEGORICAL ? val : this->feature_value);
```

Comment:
I wish we could just call `CFG::get_categorical_factor_params` here and eliminate the last `feature_value_t` parameters from the call chain?  It seems most of the necessary params are already passed down here, but I may have overlooked something important from `CFG::update_weight`.  Distributing these branches and having to look at multiple functions across multiple classes used to make it unreasonably hard to understand simple things.